### ISSUE TITLE
Storage: add an API method to give us a streaming file object

### ISSUE THREAD

**Issue by dmsolow** [googleapis/google-cloud-python#7218]
It doesn't look like there's a way to get a streaming download from google storage in the Python API. We have `download_to_file` , `download_to_string`, and `download_to_filename`, but I don't see anything that returns a file-like object that can be streamed. This is a disadvantage for many file types which can usefully be processed as they download.

Can a method like this be added?

**Comment by tseaver** [comment]
@dmsolow Hmm, `Blob.download_to_file` *takes* a file object -- does that not suit your usecase?

**Comment by dmsolow** [comment]
I don't think so. The situation is that it's often useful to start processing a file as it downloads instead of waiting until it's finished. For example if there's 1GB CSV file in google storage, it should be possible to parse it line by line as it's downloaded.

It's fairly common for network libraries to offer this kind of functionality. For example in the standard `urllib.request` HTTP library:

```
import urllib.request
import csv
from io import TextIOWrapper

with urllib.request.urlopen('http://test.com/big.csv') as f:
    wrapped = TextIOWrapper(f) # decode from bytes to str
    reader = csv.reader(wrapped)
    for row in reader:
       print(row[0])
```

This parses the CSV as it's downloaded. I'd like to get the same functionality from google storage. If there's already a good way to do this with the current library, please let me know.

**Comment by tseaver** [comment]
Hmm, you'll need to have the "stream consumer" running in a separate thread / process to do much good.  You *can* make this work using Python's `os.pipe`:  see [this gist](https://gist.github.com/tseaver/eebb9f8605f6e606b4a985e7b8ea8f85), which produces the following output:

```bash
$ bin/python pipe_test.py 
reader: start
reader: read one chunk
reader: read one chunk
...
reader: read one chunk
reader: read 800000 bytes
```

**Comment by dmsolow** [comment]
Using a separate thread kind of feels like a hack to me, but it is surely one way to do it. I think the ability to do this without using extra threads would be widely useful, but idk how hard it would be to implement.

**Comment by tseaver** [comment]
OK, looking at the [underlying implementation in `google-resumable-media`](https://github.com/googleapis/google-resumable-media-python/blob/34a36be0c53b7517bd3ddc2c281bdaaa5b9c4ef7/google/resumable_media/requests/download.py#L117-L125), all that we actually expect of the `file` object is that it has a `write` method, which is then passed each chunk as it is downloaded.

You could therefore pass in an instance of your own class which wrapped the underlying stream, e.g.:

```python
from google.cloud.storage import Client

class ChunkParser(object):

    def __init__(self, fileobj):
        self._fileobj = fileobj

    def write(self, chunk):
        self._fileobj.write(chunk)
        self._do_something_with(chunk)

client = Client()
bucket = client.get_bucket('my_bucket_name')
blob = bucket.blob('my_blob.xml')'

with open('my_blob.xml', 'wb') as blob_file:
    parser = ChunkParser(blob_file)
    blob.download_to_file(parser)
```

**Comment by yiga2** [comment]
This was requested many times but was at some point turned down (https://github.com/googleapis/google-cloud-python/issues/3903)

As an alternative, one can use the [`gcsfs`](https://github.com/dask/gcsfs) library which supports file-obj for read and write.

**Comment by dmsolow** [comment]
It's a shame that this was turned down. It's a feature that every python dev is going to expect from a library like this, as evidenced by the fact that it keeps coming up.

**Comment by akuzminsky** [comment]
> Hmm, you'll need to have the "stream consumer" running in a separate thread / process to do much good. 

Unfortunately this doesn't work with uploading streams.
https://github.com/googleapis/google-cloud-python/blob/master/storage/google/cloud/storage/blob.py#L1160 returns size of a pipe equal to zero. As result the pipe never empties and thus a child gets eventually blocked writing to it.

Are there known workarounds?

**Comment by tseaver** [comment]
@akuzminsky The line you've linked to is in the implementation of `Blob.upload_from_filename`.  This issue is about being able to process downloaded chunks before the download completes.

@dmsolow Does my file-emulating wrapper class solution work for you?

**Comment by dmsolow** [comment]
@tseaver No. I would like something that is a "file-like object." This means something that supports standard Python io methods like `readline`, `next`, `read` etc. Maybe that object buffers chunks under the hood, but it should essentially be indistinguishable from the file object returned by the builtin `open` function.

**Comment by pluvias** [comment]
@dmsolow,

> Using a separate thread kind of feels like a hack to me, but it is surely one way to do it. I think the ability to do this without using extra threads would be widely useful, but idk how hard it would be to implement.

Honestly, I am not sure, but I think it is possible and should be quite easy to use Apache Spark (or smth like this) here.

**Comment by thnee** [comment]
I was really surprised to see that not only is this feature not available, but it also has been brought up and closed in the past. It seems like an obvious and important feature to have.

Fortunately, `gcsfs` works really well as a substitute, but it's a little bit awkward to have to have a second library for such a core functionality.

But `gcsfs` does not support setting `Content-Type`, so I end up having to first upload the file using `gcsfs`, and then call `gsutil setmeta` via `subprocess` to set it after the file has been uploaded. This takes extra time and it is brittle, it is more of a workaround than a solution.

**Comment by yiga2** [comment]
@thnee you should check back, ‘gcsfs’ has the ’setxattrs()‘ method to set metadata, including content-type.

**Comment by ElliotSilver** [comment]
The lack of a simple streaming interface is a challenge to implementing a cloud function that reads/writes large files. I need the ability to read an object in from cloud storage, manipulate it, and write it out to another object. Since the only filestore available to GCF is /tmp which lives in the function memory space, you are limited to files less than 2 GB.

**Comment by IlyaFaer** [comment]
Well, if this new method is so much wanted, I'd propose solution: class, that inherits `FileIO`. It inits `ChunkedDownload` in self property and then on every `read()` call it consumes next chunk and returns it. New blob-method will be initializing this object and returning it to user (may write it to file or not)

Looks like it'll work, 'cause (as I know) most file methods works through `read()`, so overriding it must do the trick. I've already raw-coded this and tried some tests - it worked. And it's compact

**Comment by olejorgenb** [comment]
Tensorflow have an implementation that gives a file like object for gc blobs: https://www.tensorflow.org/api_docs/python/tf/io/gfile/GFile

Not sure if it actually streams or not though.

**Comment by petedannemann** [comment]
[smart_open](https://github.com/RaRe-Technologies/smart_open) now has support for streaming files from GCS.

Disclaimer: I wrote it

**Comment by rocketbitz** [comment]
@petedannemann great work - any ETA for an official release?

**Comment by petedannemann** [comment]
@rocketbitz no idea but for now you could install from Github
```bash
pip install git+https://github.com/RaRe-Technologies/smart_open
```
